src/sentry/models/releasefile.py

```diff
@@ -177,45 +194,144 @@ def extract(self) -> TemporaryDirectory:
         return temp_dir
 
 
-def merge_release_archives(file1: IO, archive2: ReleaseArchive, target: IO) -> bool:
-    """Append contents of archive2 to copy of file1
+class _Manifest:
+    def __init__(self, data: dict, fresh=False):
+        self._data = data
+        self.changed = fresh
+
+    @property
+    def data(self):
+        """Meant to be read-only"""
+        return self._data
+
+    def get(self, filename: str):
+        return self._data.get("files", {}).get(filename, None)
+
+    def update_files(self, files: dict):
+        if files:
+            self._data.setdefault("files", {}).update(files)
+            self.changed = True
+
+    def delete(self, filename: str):
+        self._data.get("files", {}).pop(filename, None)
+        self.changed = True
+
+
+class _ManifestGuard:
+    def __init__(self, release: Release, dist: Optional[Distribution]):
+        self._release = release
+        self._dist = dist
+
+    def readable_data(self) -> Optional[dict]:
+        """Simple read, no synchronization necessary"""
+        file_ = self._get_file(lock=False)
+        if file_ is not None:
+            with file_.getfile() as fp:
+                return json.load(fp)
+
+    @contextmanager
+    def writable_data(self, create: bool):
+        """Context manager for editable release manifest"""
+        with transaction.atomic():
+            if create:
+                file_, created = self._get_or_create_file()
+            else:
+                file_ = self._get_file(lock=True)
+                created = False
+
+            if file_ is None:
+                manifest = None
+            else:
+                if created:
+                    manifest = _Manifest({}, fresh=True)
+                else:
+                    data = json.load(file_.getfile())
+                    manifest = _Manifest(data)
+
+            yield manifest  # editable reference to manifest
+
+            if manifest is not None and manifest.changed:
+                file_.putfile(BytesIO(json.dumps(manifest.data).encode()))
+
+    def _get_or_create_file(self) -> Tuple[File, bool]:
+        qs = ReleaseFile.objects.select_related("file").select_for_update()
+        release_file, created = qs.get_or_create(
```

Comment:
This `get_or_create` selects with the full set of _kwargs_. Are you sure that you don't need to select with a subset of arguments and move the rest into either `defaults` or an explicit `update()`?